onnx/onnx.in.proto

```diff
@@ -196,6 +204,117 @@ message NodeProto {
+// for each stage.
+//
+// In addition to training (improving the model based on input data),
+// TrainingInfoProto can store an initialization process (resetting the model
+// back to its original state as if no training has been consumed) as a
+// computation graph in the "initialization" field. The execution of
+// "initialization" is similar to that of "algorithm". The user can use
+// "initialization_binding" to assign the outputs of "initialization" to
+// ModelProto.graph.initializer and TrainingInfoProto.algorithm.initializer.
+// 
+// Several inference components are usable in TrainingInfoProto; the list
+// includes
+//   1. ModelProto.graph is a callable inside TrainingInfoProto.algorithm.
+//      Each call is a NodeProto in TrainingInfoProto.algorithm.node with
+//      NodeProto.op_type == "Graph" and
+//      NodeProto.name == ModelProto.graph.name.
+//   2. Initializers in ModelProto.graph.initializer can be referenced by
+//      nodes in TrainingInfoProto.algorithm.nodes and
+//      TrainingInfoProto.initialization.nodes. It implies that training
+//      nodes cannot produce values with names identical to inference
+//      initializers.
+message TrainingInfoProto {
+  // This field describes a graph to compute the initial tensors
+  // upon starting the training process. Initialization graph has no input
+  // and can have multiple outputs. Usually, trainable tensors in neural
+  // networks are randomly initialized. To achieve that, for each tensor,
+  // the user can put a random operator such as RandomNormal or
+  // RandomUniform in TrainingInfoProto.initialization.node and assign its
+  // random output to the specific tensor using "initialization_binding".
+  // This graph can also set the initializers in "algorithm"; a use case is
+  // resetting the number of training iteration to zero.
+  //
+  // By default, this field is an empty graph and its evaluation does not
+  // produce any output.
+  optional GraphProto initialization = 1;
+
+  // This field represents a training algorithm step. Given required inputs,
+  // it computes outputs to update initializers in its own initializer
+  // list and ModelProto.graph.initializer. In general, this graph contains
+  // loss node, gradient node, optimizer node, increment of iteration count,
+  // and calls to the inference graph.
+  //
+  // The field algorithm.node can invoke ModelProto.graph as a node.
+  // The initializers defined in ModelProto.graph.initializer are visible in
+  // the "algorithm".
+  //
+  // The field MUST be present.
+  optional GraphProto algorithm = 2;
+
+  // This field specifies the bindings from the outputs of "initialization" to
+  // the initializers in "algorithm" and "ModelProto.graph". See
+  // "update_binding" below for its format.
+  //
+  // By default, this field is empty and no initializer would be changed
+  // by the execution of "initialization".
+  optional StringStringEntryProto initialization_binding = 3;
```

Comment:
Thanks for the fixing!

Question: do we need this explicit name binding/mapping? In my opinion. using the same name string should be good enough, instead of adding one more mapping table. Thoughts?

Reply:
It may break SSA in a graph. If inside the training algorithm graph, we store the following two equations to update `W`:
```
W, X->Sub->W
W, Y->Add->Z
```
where `W`, `X`, `Y`, and `Z` are tensors. Should `W, Y -> Add -> Z` use the output `W` of `Sub`?

Reply:
fair.